pkg/plugin/query.go

```diff
@@ -119,13 +124,71 @@ func (q *Query) queryRangeURL(expr string, step time.Duration, queryParams url.V
 	return q.url.String()
 }
 
-var legendReplacer = strings.NewReplacer("{{", "", "}}", "")
+var legendReplacer = regexp.MustCompile(`\{\{\s*(.+?)\s*\}\}`)
 
-func (q *Query) parseLegend() string {
-	legend := legendReplacer.Replace(q.LegendFormat)
+func (q *Query) parseLegend(field *data.Field) string {
+	labels := field.Labels
+	legend := metricNameFromLabels(field)
 
+	if q.LegendFormat == legendFormatAuto {
+		if len(labels) > 0 {
+			legend = ""
+		}
+	} else if q.LegendFormat != "" {
+		result := legendReplacer.ReplaceAllStringFunc(q.LegendFormat, func(in string) string {
+			labelName := strings.Replace(in, "{{", "", 1)
+			labelName = strings.Replace(labelName, "}}", "", 1)
+			labelName = strings.TrimSpace(labelName)
+			if val, exists := labels[labelName]; exists {
+				return val
+			}
+			return ""
+		})
+		legend = result
+	}
+
+	// If legend is empty brackets, use query expression
 	if legend == "{}" {
 		return q.Expr
 	}
+
 	return legend
 }
+
+func (q *Query) addMetadataToMultiFrame(frame *data.Frame) {
+	if len(frame.Fields) < 2 {
+		return
+	}
+
+	customName := q.parseLegend(frame.Fields[1])
+	if customName != "" {
+		frame.Fields[1].Config = &data.FieldConfig{DisplayNameFromDS: customName}
+	}
+
+	frame.Name = customName
+}
+
+func metricNameFromLabels(f *data.Field) string {
```

Comment:
This function is also confusing... The lack of description or tests makes me take guesses. And I get confused of `numLabels` usage...

Reply:
renamed function made it simpler, added tests